### PR TITLE
ENH/CLN: Fix DataType hashing

### PR DESCRIPTION
**Pull request by cpcloud**
Before this PR, every datatype's hash function was `hash(type(self))`, which
means collisions galore (triggering __eq__ comparison when using complex types
as dict keys). This PR provides a hash function based on ``__slots__`` giving
every type a unique hash based on ``type(self)``, its slots (if any) + the
``nullable`` field from the ``DataType`` parent class.

### CODE REVIEW

**Comment by cpcloud** [comment]
@wesm can you review this when you get a chance?

File: ibis/expr/datatypes.py [hunk 1]
```diff
@@ -147,42 +147,35 @@ class HasSchema(object):
...
 
+    __slots__ = 'schema', 'name'
+
     def __init__(self, schema, name=None):
-        assert isinstance(schema, Schema)
-        self._schema = schema
-        self._name = name
+        if not isinstance(schema, Schema):
+            raise TypeError(
+                'schema argument to HasSchem class must be a Schema instance'
```
Comment (wesm) [inline ibis/expr/datatypes.py]: Minor typo

File: ibis/expr/datatypes.py [hunk 2]
```diff
@@ -390,78 +415,79 @@ def valid_literal(self, value):
 
 class Int8(Integer):
 
+    __slots__ = ()
```
Comment (wesm) [inline ibis/expr/datatypes.py]: Do `__slots__` inherit in any reasonable way? (I assume not)

File: ibis/expr/datatypes.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/datatypes.py]: Thanks.

File: ibis/expr/datatypes.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/datatypes.py]: That's right. Stated here: https://docs.python.org/3.6/reference/datamodel.html#notes-on-using-slots
> The action of a `__slots__` declaration is limited to the class where it is defined. As a result, subclasses will have a `__dict__` unless they also define `__slots__` (which must only contain names of any additional slots).

**Comment by cpcloud** [comment]
Merging on green.

**Pull request closed without merging**